### ISSUE TITLE
E2E 1.11.1 does not work properly and is automatically selected

### ISSUE THREAD

**Issue by MSe1969** [nextcloud/end_to_end_encryption#374]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.

### Steps to reproduce
1. Fresh install E2E in Nextcloud 25.0.1, OR
2. Upgrade from NC24 to NC25 while having E2E installed and in use

### Expected behaviour
Tell us what should happen
According to the discussion in #357 version 1.11.0 is the proper stable version, but 1.11.1 is being installed or updated as app.
The automatically taken version is supposed to work, this means:
In an encrypted directory on the server, the file names are hash codes and the file content is encrypted. On the attached sync clients, the files are synchronized and decrypted, so a file created in one client, is encrypted and uploaded to the server in the a.m. encrypted state. On a further sync client, the file is downloaded and decrypted and appears usable normally.

### Actual behaviour
Tell us what happens instead
E2E Version 1.11.1 does not work properly:
If a new file is added to the local directory having the "encrypted" flag, it is uploaded UNENCRYPTED to the server and appears in the encrypted directory unencrypted with its original file name. On any attached further sync client (2nd PC/laptop in my case), an error message appears to indicate, that the file could not be synchronized.
The same also happens, if a new directory is created and flagged for encryption.

So, either configure the app directory to install E2E 1.11.0, if that is supposed to work properly, or fix 1.11.1

### Server configuration

**Operating system:**
Debian 11

**Web server:**
Nginx 1.23.2

**Database:**
mysql  Ver 15.1 Distrib 10.8.6-MariaDB

**PHP version:**
PHP 8.1.13

**Nextcloud version:** (see Nextcloud admin page)
25.0.1

**Updated from an older Nextcloud/ownCloud or fresh install:**
Update from 24.0.7

**Where did you install Nextcloud from:**
https://download.nextcloud.com/server/releases/latest.tar.bz2

```
Login as admin user into your Nextcloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results here.
```
No errors have been found.

**List of activated apps:**
```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your Nextcloud installation folder

If not, provide at least the version number of the End-to-End-Encryption app.
```
Enabled:
  - activity: 2.17.0
  - admin_audit: 1.15.0
  - circles: 25.0.0
  - cloud_federation_api: 1.8.0
  - comments: 1.15.0
  - contactsinteraction: 1.6.0
  - dashboard: 7.5.0
  - dav: 1.24.0
  - end_to_end_encryption: 1.11.1
  - federatedfilesharing: 1.15.0
  - federation: 1.15.0
  - files: 1.20.1
  - files_external: 1.17.0
  - files_pdfviewer: 2.6.0
  - files_rightclick: 1.4.0
  - files_sharing: 1.17.0
  - files_trashbin: 1.15.0
  - files_versions: 1.18.0
  - logreader: 2.10.0
  - lookup_server_connector: 1.13.0
  - nextcloud_announcements: 1.14.0
  - notifications: 2.13.1
  - oauth2: 1.13.0
  - password_policy: 1.15.0
  - photos: 2.0.0
  - privacy: 1.9.0
  - provisioning_api: 1.15.0
  - recommendations: 1.4.0
  - related_resources: 1.0.3
  - richdocuments: 7.0.2
  - richdocumentscode: 22.5.802
  - serverinfo: 1.15.0
  - settings: 1.7.0
  - sharebymail: 1.15.0
  - support: 1.8.0
  - systemtags: 1.15.0
  - text: 3.6.0
  - theming: 2.0.1
  - twofactor_backupcodes: 1.14.0
  - updatenotification: 1.15.0
  - user_status: 1.5.0
  - viewer: 1.9.0
  - weather_status: 1.5.0
  - workflowengine: 2.7.0
Disabled:
  - bruteforcesettings
  - encryption
  - firstrunwizard: 2.13.0
  - survey_client: 1.12.0
  - suspicious_login
  - twofactor_totp
  - user_ldap

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your Nextcloud installation folder
```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "basar.pmvl.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "25.0.1.1",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "activity_expire_days": 14,
        "allow_local_remote_servers": true,
        "auth.bruteforce.protection.enabled": true,
        "blacklisted_files": [
            ".htaccess",
            "Thumbs.db",
            "thumbs.db"
        ],
        "cron_log": true,
        "default_phone_region": "DE",
        "defaultapp": "files,dashboard",
        "enable_previews": true,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\Movie",
            "OC\\Preview\\PDF",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown"
        ],
        "filesystem_check_changes": 0,
        "filelocking.enabled": "true",
        "htaccess.RewriteBase": "\/",
        "integrity.check.disabled": false,
        "knowledgebaseenabled": false,
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "loglevel": 2,
        "logtimezone": "Europe\/Berlin",
        "log_rotate_size": "104857600",
        "maintenance": false,
        "maintenance_window_start": 1,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "overwriteprotocol": "https",
        "preview_max_x": 1024,
        "preview_max_y": 768,
        "preview_max_scale_factor": 1,
        "profile.enabled": false,
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0.5,
            "dbindex": 1
        },
        "quota_include_external_storage": false,
        "share_folder": "\/Freigaben",
        "skeletondirectory": "",
        "theme": "",
        "trashbin_retention_obligation": "auto, 7",
        "updater.release.channel": "stable"
    }
}

</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
no

**Are you using encryption:** yes/no
no sever-side encryption, but E2E app

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

### Client configuration
**Browser:** Firefox

**Operating system:**
Linux Debian 11

#### Web server error log
n/a

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
Insert your Nextcloud log here
```
</details>

#### Browser log
n/a

**Comment by eibex** [comment]
On iOS new files are encrypted properly. I will test this later today on desktop.

The issue may be related to the desktop app instead of the e2ee api.

**Comment by MSe1969** [comment]
Updating NC to 25.0.2 hasn't changed the situation; (1.11.1 is meanwhile the "official" version).
Upgrading the Linux client hasn't brought any improvement either. **It does not work**, as described above - files are uploaded in an unencrypted state.

I see in the E2E code, that the method to check, whether a directory is encrypted has changed.
Could that be the reason?

Any chance to downgrade the app ? The 1.10 version was properly working, also in NC25 (after adapting info.xml) but downgrading the app (and editing info.xml again) does not work at all ...

**Comment by MSe1969** [comment]
>     * checked storage on server (`/.../Username/files/Encrypted/*`): everything encrypted
> 
> Probably not helpful, but at least the message: it is not failing for everyone.

I have meanwhile tested this in different setups : 

Clients: 
1. Windows (2019 Server, AWS instance), 
2. Ubuntu 22.04 KDE (laptop), 
3. Debian 11 Cinnamon (desktop PC)

Server: 
1. My above setup, 
2. Temp. "1-click NC setup" on Hetzner cloud
3. NC 25 installation on Web-Hosting package (Manitu)

The result is **ALWAYS** the same, since my first tests with the NC 25 beta:
It does not matter, whether an existing directory with already encrypted data is used or a fresh directory is created and encrypted, it does not matter whether an NC server is upgraded for freshly installed.
- The 1st uploaded new file is encrypted OK, 
- every further uploaded file is not encrypted any more
- All already existing encrypted files can normally be synch'ed

Please see below an example of an encrypted directory of a fresh server installation, where simply a new encrypted directory on the client was created, and four random files were pasted into:

```
# ls -la
total 1552
drwxr-xr-x 2 www-data www-data    4096 Dec 31 16:45 ./
drwxr-xr-x 7 www-data www-data    4096 Dec 31 16:44 ../
-rw-r--r-- 1 www-data www-data    1516 Apr 17  2021 10101479031618587882-128.png
-rw-r--r-- 1 www-data www-data  127448 Jan 23  2021 20190116-4-0238-11-allgemeine-bedingungen-data.pdf
-rw-r--r-- 1 www-data www-data     194 Apr  4  2021 20210404_AWS_IAM_user_credentials.csv
-rw-r--r-- 1 www-data www-data 1439249 Apr  2  2022 87032a86c7c74eae80c78d395a8af2c3
```

The issue can easily be overlooked and going unnoticed, if a test cycle simply consists of one single file.
Nevertheless, based on the testing, which I have performed, there is an issue with the E2E app, this for sure is not related to my personal setup.

**Comment by bcutter** [comment]
Very interesting. Here's a random encrypted folder:

![image](https://user-images.githubusercontent.com/13799156/210152915-94c09455-aebe-4f21-b44c-423ba18e2bcb.jpeg)

So I put m bets on the initialization of the E2EE. Probably something with the key Management is buggy, that could clarify why all new users (fresh install of newer version) like you suffer from this, while existing/long-term users upgrading from earlier versions have no issues.

Without an expert looking into the code... not much to expect.

**Comment by MSe1969** [comment]
> On iOS new files are encrypted properly. I will test this later today on desktop.
> 
> The issue may be related to the desktop app instead of the e2ee api.

Well, seems a combination of several factory playing together - I think I finally found the solution:

**The good news first:**
I have downloaded the daily client build on 3.6.5 from today: 

- linux-20230101.AppImage on my Linux laptop 

- windows-20230101.msi on my Windows VM

And now, all perfectly works.


What is worth mentioning (see also linked issue) #357 is also the following observation:

1. Using Client 3.6.0 or 3.6.3 or 3.6.4 with NC25 and E2E 1.11.x => fail, as described
2. Using Client 3.6.3 / 3.6.4 with NC25 and E2E 1.10.0 and edited info.xml (**before** 1.11.1 was officially released) => OK
3. Once E2E 1.11.1 was officially released, a forced update took place and replacing the E2E app with 1.10.0 by simultaneously updating info.xml with NC 25 and app version 1.11.4 to prevent forced updates has lead to the same misbehavior => fail
4. Client 3.6.5 with NC25 and E2E 1.11.1 => works OK


It seems, that E2E version 1.11.x requires a different handling initiated by the client. And the 1.11.x handling was faulty and only fixed in Desktop/client 3.6.5 - the commit https://github.com/nextcloud/desktop/commit/bf1f718a6561a496f7de259269fa9663d241b3f4 seems to play a major role in the fix:

Quote:
```
    // Folder records have directory paths in Foo/Bar/ convention...
    // But EncryptFolderJob expects directory path Foo/Bar convention
```

So my issue is solved now, but this should somehow lead to an accelerated deployment of the next Desktop release. I'll open an issue there.
